src/webgpu/api/validation/encoding/cmds/render/dynamic_state.spec.ts

```diff
@@ -141,60 +142,89 @@ TODO Test the first value smaller than -0`
 
-    // Test -1
-    { x: -1, y: 0, w: 0, h: 0 },
-    { x: 0, y: -1, w: 0, h: 0 },
+    // Negative width/height is invalid
     { x: 0, y: 0, w: -1, h: 0 },
     { x: 0, y: 0, w: 0, h: -1 },
+
+    // Negative width/height is invalid even if the resulting bounds are positive
+    { x: 1, y: 0, w: -1, h: 0 },
+    { x: 0, y: 1, w: 0, h: -1 },
   ])
   .fn(t => {
     const { x, y, w, h } = t.params;
-    const success = x >= 0 && y >= 0 && w >= 0 && h >= 0;
+    const success = w >= 0 && h >= 0;
     t.testViewportCall(success, { x, y, w, h, minDepth: 0, maxDepth: 1 });
   });
 
-g.test('setViewport,xy_rect_contained_in_attachment')
+g.test('setViewport,exceeds_attachment_size')
+  .desc(`Test that the viewport can exceed the attachment size`)
+  .paramsSubcasesOnly([
+    { attachmentWidth: 3, attachmentHeight: 3 },
+    { attachmentWidth: 1024, attachmentHeight: 1024 },
+  ])
+  .fn(t => {
+    const { attachmentWidth, attachmentHeight } = t.params;
+    t.testViewportCall(
+      true,
+      { x: 0, y: 0, w: attachmentWidth + 1, h: attachmentHeight + 1, minDepth: 0, maxDepth: 1 },
+      { width: attachmentWidth, height: attachmentHeight, depthOrArrayLayers: 1 }
+    );
+  });
+
+g.test('setViewport,xy_rect_contained_in_bounds')
   .desc(
-    'Test that the rectangle defined by x, y, width, height must be contained in the attachments'
+    `Test that the rectangle defined by x, y, width, height must be contained in the maximum viewport bounds
+and that the viewport size cannot exceed the maximum.`
   )
   .paramsSubcasesOnly(u =>
-    u
-      .combineWithParams([
-        { attachmentWidth: 3, attachmentHeight: 5 },
-        { attachmentWidth: 5, attachmentHeight: 3 },
-        { attachmentWidth: 1024, attachmentHeight: 1 },
-        { attachmentWidth: 1, attachmentHeight: 1024 },
-      ])
-      .combineWithParams([
-        // Control case: a full viewport is valid.
-        { dx: 0, dy: 0, dw: 0, dh: 0 },
-
-        // Other valid cases with a partial viewport.
-        { dx: 1, dy: 0, dw: -1, dh: 0 },
-        { dx: 0, dy: 1, dw: 0, dh: -1 },
-        { dx: 0, dy: 0, dw: -1, dh: 0 },
-        { dx: 0, dy: 0, dw: 0, dh: -1 },
-
-        // Test with a small value that causes the viewport to go outside the attachment.
-        { dx: 1, dy: 0, dw: 0, dh: 0 },
-        { dx: 0, dy: 1, dw: 0, dh: 0 },
-        { dx: 0, dy: 0, dw: 1, dh: 0 },
-        { dx: 0, dy: 0, dw: 0, dh: 1 },
-      ])
+    u.combine('dimension', [0, 1]).combineWithParams([
+      // Control case: max viewport is valid.
+      { om: 0, od: 0, sd: 0 },
+
+      // Other valid cases
+      { om: -1, od: 0, sd: 0 },
+      { om: -2, od: 0, sd: 0 },
+      { om: 1, od: -1, sd: 0 },
+      { om: 0, od: -1, sd: 0 },
+      { om: 0, od: 1, sd: 0 },
+      { om: 1, od: 0, sd: -1 },
+
+      // Cases that go outside the allowed bounds
+      { om: -2, od: -1, sd: 0 },
+      { om: 1, od: 0, sd: 0 },
+      { om: 1, od: 1, sd: -1 },
+      { om: 1, od: -0.1, sd: 0 },
```

Comment:
For each of these we should ideally test two cases that are just an epsilon apart, one valid and one invalid. (see `nextAfterF32` in `math.ts`)

and in order to make it clear that those test cases are actually doing what we expect, it would help to put `_success` as a parameter, instead of computing it in the test